File: backend/schemas/signal.py
"""信号相关 Pydantic Schema

包含：SignalOut, AlertSettingIn, LiveSignalFilter, LiveSignalStatsOut, EmotionDataOut, EmotionFlowDataOut
"""

from datetime import datetime

from pydantic import BaseModel, Field


class SignalOut(BaseModel):
    """信号详情输出"""

    id: str
    stock_code: str
    stock_name: str | None = None
    signal_type: str
    strategy_id: str | None = None
    signal_time: datetime
    signal_price: float = 0.0
    mode: str = "simulated"
    confidence: float = 0.0
    status: str = "pending"
    stop_loss_price: float | None = None
    take_profit_price: float | None = None
    alert_status: str = "none"
    actual_pnl: float | None = None
    holding_days: int | None = Field(None, description="持仓天数（计算字段）")
    current_price: float | None = Field(None, description="当前价格（计算字段）")
    floating_pnl: float | None = Field(None, description="浮动盈亏百分比（计算字段）")

    model_config = {"from_attributes": True}


class AlertSettingIn(BaseModel):
    """预警设置输入"""

    stop_loss_price: float | None = Field(None, description="止损价")
    take_profit_price: float | None = Field(None, description="止盈价")
    alert_method: str = Field("push", description="预警方式: push | sms | email")
    alert_frequency: str = Field("realtime", description="预警频率: realtime | daily | weekly")


class LiveSignalFilter(BaseModel):
    """实时信号筛选条件"""

    type: str | None = Field(None, description="信号类型筛选")
    period: str | None = Field(None, description="时间周期: today | week | month")
    search: str | None = Field(None, description="股票代码/名称搜索")
    onlyHolding: bool = Field(False, description="仅显示持仓中")
    onlyAlerting: bool = Field(False, description="仅显示预警中")


class LiveSignalStatsOut(BaseModel):
    """实时信号统计"""

    today_count: int = 0
    holding: int = 0
    take_profit: int = 0
    stop_loss: int = 0
    alerting: int = 0


class EmotionDataOut(BaseModel):
    """市场情绪数据点"""

    date: str
    value: float = 0.0


class EmotionFlowDataOut(BaseModel):
    """市场情绪资金流数据点"""

    date: str
    inflow: float = 0.0
    outflow: float = 0.0

File: src/types/signal.ts
import type { SignalTagType } from './stock';

/** 信号类型 */
export type SignalType = 'BUY' | 'SELL';

/** 信号状态 */
export type SignalStatus = 'pending' | 'executed' | 'ignored' | 'expired' | 'closed';

/** 预警状态 */
export type AlertStatus = 'safe' | 'warning' | 'stop_loss' | 'take_profit';

/** 交易信号 */
export interface Signal {
  id: string;
  stock_code: string;
  stock_name: string;
  signal_type: SignalType;
  strategy_id: string;
  signal_time: string;
  signal_price: number;
  mode: SignalTagType;
  confidence: number;
  status: SignalStatus;
  stop_loss_price: number | null;
  take_profit_price: number | null;
  alert_status: AlertStatus;
  actual_pnl: number | null;
  holding_days?: number;
  current_price?: number;
  floating_pnl?: number;
}

/** 实盘信号筛选 */
export interface LiveSignalFilter {
  type: 'all' | 'BUY' | 'SELL';
  period: 'today' | 'week' | 'month';
  search?: string;
  onlyHolding?: boolean;
  onlyAlerting?: boolean;
}

/** 实盘信号统计 */
export interface LiveSignalStats {
  today_count: number;
  holding: number;
  take_profit: number;
  stop_loss: number;
  alerting: number;
}

/** 预警设置 */
export interface AlertSetting {
  stop_loss_price: number;
  take_profit_price: number;
  alert_method: 'notification' | 'sms' | 'wechat';
  alert_frequency: 'once' | 'daily' | 'every';
}

/** 择时信号映射 */
export interface TimingSignalMap {
  [mode: string]: Signal[];
}
